f3/certstore.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	return cs, nil
+}
+
+var certStoreLatestKey = datastore.NewKey("/latestCert")
+
+func (cs *CertStore) loadLatest(ctx context.Context) (*Cert, error) {
+	cb, err := cs.ds.Get(ctx, certStoreLatestKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("getting latest cert: %w", err)
+	}
+	var c Cert
+	err = c.UnmarshalBinary(cb)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshalling latest cert: %w", err)
+	}
+	return &c, err
+}
+
+// Latest returns the newest available certificate
+func (cs *CertStore) Latest() *Cert {
+	return cs.busCerts.Last()
+}
+
+func (cs *CertStore) Get(ctx context.Context, instance uint64) (*Cert, error) {
+	b, err := cs.ds.Get(ctx, cs.keyForInstance(instance))
+
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, xerrors.Errorf("cert at %d: %w", instance, ErrCertNotFound)
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("accessing cert in datastore: %w", err)
+	}
+
+	var c Cert
+	if err := c.UnmarshalBinary(b); err != nil {
+		return nil, xerrors.Errorf("unmarshalling cert: %w", err)
+	}
+	return &c, err
+}
+
+// GetRange returns a range of certs from start to end inclusive by instance numbers in the
+// increasing order. Only this order of traversal is supported.
+// If it encounters missing cert, it returns a wrapped ErrCertNotFound and the available certs
+func (cs *CertStore) GetRange(ctx context.Context, start uint64, end uint64) ([]Cert, error) {
+	if start > end {
+		return nil, xerrors.Errorf("start is larger then end: %d > %d", start, end)
+	}
+	if end-start > uint64(math.MaxInt)-1 {
+		return nil, xerrors.Errorf("range %d to %d is too large", start, end)
+	}
+
+	bCerts := make([][]byte, 0, end-start+1)
+
+	for i := start; i <= end; i++ {
+		b, err := cs.ds.Get(ctx, cs.keyForInstance(i))
+		if errors.Is(err, datastore.ErrNotFound) {
+			break
+		}
+		if err != nil {
+			return nil, xerrors.Errorf("accessing cert at %d for range request: %w", i, err)
+		}
+
+		bCerts = append(bCerts, b)
+	}
+
+	certs := make([]Cert, len(bCerts))
+	for j, bCert := range bCerts {
+		err := certs[j].UnmarshalBinary(bCert)
+		if err != nil {
+			return nil, xerrors.Errorf("unmarshalling a cert at j=%d, instance %d: %w", j, start+uint64(j), err)
+		}
+	}
+
+	if len(certs) < cap(bCerts) {
+		return certs, xerrors.Errorf("cert at %d: %w", start+uint64(len(bCerts)), ErrCertNotFound)
+	}
+	return certs, nil
+}
+
+func (_ *CertStore) keyForInstance(i uint64) datastore.Key {
+	return datastore.NewKey(fmt.Sprintf("/certs/%016X", i))
+}
+
+// Put saves a certificate in a store and notifies listeners.
+// It errors if the certificate is already in the store
+func (cs *CertStore) Put(ctx context.Context, cert *Cert) error {
+	key := cs.keyForInstance(cert.Instance)
+
+	exists, err := cs.ds.Has(ctx, key)
+	if err != nil {
+		return xerrors.Errorf("checking existence of cert: %w", err)
+	}
+	if exists {
+		return ErrCertAlreadyExists
```

Comment:
nit: any reason to return an error rather than making this idempotent?

Reply:
Also, not thread safe.

Reply:
You are right that making it idempotent is a better pattern